### PR TITLE
sync and write containership ssh keys

### PR DESCRIPTION
**Pull request by mikestoltz**
@normanjoyner @nicktate

### CODE REVIEW

File: lib/follower.js [hunk 1]
```diff
@@ -16,6 +17,9 @@ module.exports = {
     initialize: function(core, config) {
         core.cluster.legiond.join('containership-cloud.backup');
 
+        // add ssh keys
```
Comment (normanjoyner) [inline lib/follower.js]: `// enable SSH key management`

File: lib/leader.js [hunk 2]
```diff
@@ -16,6 +17,7 @@ module.exports = {
     initialize: function(core, config) {
         const firewall = new Firewall(core);
         firewall.enable();
+        new SSH_Keys(core);
```
Comment (normanjoyner) [inline lib/leader.js]: `// enable SSH key management`

File: lib/ssh_keys.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+            const subscriber = this.myriad_client.subscribe(constants.myriad.SSH_KEYS_REGEX);
+
+            subscriber.on('message', (message) => {
+                if(message.type === 'data') {
+                    this.write_keys();
+                }
+            });
+        } else {
+            this.core.loggers['containership-cloud'].log('error', 'Plugin requires vesrion 1.8.0 of containership or greater! Refusing to set ssh keys!');
```
Comment (normanjoyner) [inline lib/ssh_keys.js]: `version`* this was a typo in `lib/firewall.js`

File: lib/ssh_keys.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    write_keys() {
+        this.core.loggers['containership-cloud'].log('verbose', 'Updating ssh keys');
+
+        this.get_keys((err, keys) => {
+            if(err) {
+                this.core.loggers['containership-cloud'].log('error', `Error fetching containership cloud ssh keys: ${err.message}`);
```
Comment (normanjoyner) [inline lib/ssh_keys.js]: does it make sense to short-circuit here? We shouldn't be able to write anything if this is the case

File: lib/ssh_keys.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            }
+
+            try {
+                if(!fs.existsSync('/home/containership/.ssh')) {
+                    fs.mkdirSync('/home/containership/.ssh');
```
Comment (normanjoyner) [inline lib/ssh_keys.js]: check for error?

File: lib/ssh_keys.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            } catch (e) {
+                this.core.loggers['containership-cloud'].log('error', JSON.stringify(e));
+            }
+        });
+    }
+
+    get_keys(callback) {
+        this.core.cluster.myriad.persistence.get(constants.myriad.SSH_KEYS, (err, keys) => {
+            if(err) {
+                return callback(err, []);
```
Comment (normanjoyner) [inline lib/ssh_keys.js]: this seems potentially harmful as it will unexpectedly wipe existing keys in the case of failed myriad reads

Comment (normanjoyner) [inline lib/ssh_keys.js]: (will be fixed with short circuit comment above)

File: lib/ssh_keys.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            } else {
+                try {
+                    keys = JSON.parse(keys);
+                    return callback(null, keys);
+                } catch(err) {
+                    return callback(err, []);
```
Comment (normanjoyner) [inline lib/ssh_keys.js]: same case as above ... this case being less likely to occur. also resolved by short-circuit

File: lib/ssh_keys.js (hunk 3 above)
Comment (normanjoyner) [inline lib/ssh_keys.js]: (can you fix there as well please)

**Comment by mikestoltz** [comment]
@normanjoyner updated

**Pull request merged**